src/lib/traceCall.js

```diff
@@ -0,0 +1,199 @@
+// @flow
+import { EthRpc, convert } from 'emerald-js';
+import BigNumber from 'bignumber.js';
+
+type Call = {
+    method: string,
+    params: Array<any>,
+}
+
+type Transaction = {
+    from: string,
+    to: string,
+    gas: string,
+    gasPrice: string,
+    value: string,
+    data: string,
+}
+
+interface ITracer {
+    buildRequest(): Call;
+    estimateGas(trace: any): ?BigNumber;
+}
+
+/**
+ * Every Ethereum implementation has eth_estimateGas
+ */
+class CommonCallTracer implements ITracer {
+    tx: Transaction;
+
+    constructor(tx: Transaction) {
+        this.tx = tx;
+    }
+
+    buildRequest(): Call {
+        const params = [{
+            from: this.tx.from,
+            to: this.tx.to,
+            gas: this.tx.gas,
+            gasPrice: this.tx.gasPrice,
+            value: this.tx.value,
+            data: this.tx.data,
+        }];
+        params.push('latest');
+
+        return {
+            method: 'eth_estimateGas',
+            params,
+        };
+    }
+
+    estimateGas(trace: any): BigNumber | null {
+        return convert.toBigNumber(trace);
+    }
+}
+
+/**
+ * Parity has trace_call API method
+ */
+export class ParityTracer implements ITracer {
```

Comment:
this is never used?

Reply:
If I understood correctly `trace_call` doesn't evaluate gas amount needed for TX, it just trace tx according to provided gas and can just report with `error: out of gas`.